JENKINS-21336: ADMINISTER should not imply RUN_SCRIPTS

jvz opened the pull request:
This introduces a new feature flag, `hudson.security.AdministerImpliesRunScripts` which is set to true by default. Setting this to false changes the default permission behavior to use RUN_SCRIPTS as the root level permission.

This also introduces a declarative permission registry so that plugins may register permission instances to help alleviate problems such as in JENKINS-17200.

See [JENKINS-21336](https://issues.jenkins-ci.org/browse/JENKINS-21336).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Added feature flag `hudson.security.AdministerImpliesRunScripts` which can be set to `false` to disable the default legacy behavior of the `Jenkins.ADMINISTER` permission implying `Jenkins.RUN_SCRIPTS`, `PluginManager.CONFIGURE_UPDATECENTER`, and `PluginManager.UPLOAD_PLUGINS`.
* `Internal:` added `@GlobalPermissionGroup` and `@GlobalPermission` annotations to declare permission instances in static fields. These are now lazily loaded in `Permission.getAll()`, `PermissionGroup.getAll()`, etc.
* `Internal:` added `Jenkins.ROOT` permission alias to point to either `Jenkins.ADMINISTER` or `Jenkins.RUN_SCRIPTS` depending on the value of `hudson.security.AdministerImpliesRunScripts`.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @daniel-beck @Wadeck @jglick @oleg-nenashev 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

jvz committed:
JENKINS-21336: ADMINISTER should not imply RUN_SCRIPTS

This introduces a new feature flag, hudson.security.Permission.administerImpliesRunScripts which is set to true by default. Setting this to false changes the default permission behavior to use RUN_SCRIPTS as the root level permission.

This also introduces a separated permission registry so that plugins may eagerly register permission instances on startup to help alleviate problems such as in JENKINS-17200.

Also relates to internal issue CORE-1068.

jvz committed:
Make Permission Comparable

This fixes an unserializable lambda exception.

jglick commented:
This is likely going to require changes in various `AuthorizationStrategy` plugins touched by SECURITY-410.

fcojfernandez reviewed (approved):
Code looks good to me, yet I haven't checked it locally

jvz commented:
@jglick yes that makes sense. I was thinking this feature can be gradually introduced via feature flag, then when the plugins are updated to work with this, the feature flag can be toggled on by default instead. I'm not sure if we already have some sort of feature flag system, but that would certainly be useful here.

jvz committed:
Remove obsolete tests

Permissions and PermissionGroups are registered into a threadsafe data structure, so any races to register can be safely ignored.

jvz committed:
Ensure implication relationship of admin permissions works with either feature flag

jvz committed:
Fix tests around ADMINISTER to use ROOT instead

jglick commented on core/src/main/java/hudson/security/Permission.java:
```diff
@@ -274,22 +233,52 @@ public void setEnabled(boolean enable) {
+     * @return true when ADMINISTER implies RUN_SCRIPTS, false when RUN_SCRIPTS implies ADMINISTER
+     * @since TODO
+     */
+    public static boolean administerImpliesRunScripts() {
```
`@Restricted(NoExternalUse.class)` and delete `@since`, assuming there is no use case for calling this from plugins.

jvz replied:
Can do.

jglick commented on core/src/main/java/hudson/security/Permission.java:
```diff
@@ -315,7 +306,26 @@ public boolean getEnabled() {
-    public static final Permission HUDSON_ADMINISTER = new Permission(HUDSON_PERMISSIONS,"Administer", hudson.model.Messages._Hudson_AdministerPermission_Description(),null);
+    public static final Permission HUDSON_ADMINISTER;
+
+    @Restricted(NoExternalUse.class)
```
Would you not want plugins to be checking this?

jvz replied:
I marked this restricted because the real instance is in `Jenkins.ROOT` to mirror `Jenkins.ADMINISTER`. The same initialization ordering constraints apply here, hence the permission is defined in `Permission`.

jglick commented on core/src/main/java/hudson/security/Permission.java:
```diff
@@ -315,7 +306,26 @@ public boolean getEnabled() {
+    public static final Permission ROOT;
+
+    static {
+        final Permission administer, runScripts, root;
```
Could assign directly to the actual constants I think.

jvz replied:
True. I'll clean this up.

jglick commented on core/src/main/java/hudson/security/Permission.java:
```diff
@@ -274,22 +233,52 @@ public void setEnabled(boolean enable) {
+     * Comparator that orders {@link Permission} objects based on their ID.
+     */
+    @Deprecated
+    public static final Comparator<Permission> ID_COMPARATOR = Comparator.comparing(Permission::getId);
```
or just the natural ordering

jvz replied:
Oh right, I forgot that I added Comparable to Permission. That was mainly due to a JEP-200 issue with serializing lambdas.

jglick commented on core/src/main/java/hudson/security/PermissionRegistry.java:
```diff
@@ -0,0 +1,85 @@
+ * @since TODO
+ */
+@ParametersAreNonnullByDefault
+public enum PermissionRegistry {
```
Nonstandard way of making a singleton.

Also you probably need to make this instance be tied to `Jenkins.get()`, so that for example tests using `RestartableJenkinsRule` will properly get a new registry after the restart. In general, you cannot use mutable `static` fields (which this effectively is) anywhere in Jenkins other than `Jenkins` itself.

(yes this may already have been broken in `PermissionGroup`, but if so, now is the time to fix it)

jvz replied:
Nonstandard way of doing a singleton in Jenkins, perhaps, but an `enum` is the most effective singleton in Java due to guarantees by Java that it will be a true singleton regardless of multithreading, initialization ordering, etc.

I remember seeing an error when trying to use an extension point instead for the permission registry, but I don't recall what it was. I'll try that approach again and get back to you.

jvz replied:
I tried making this an `@Extension` instead of an enum, but now I get an error in the clinit of Jenkins:

```
Oct 25, 2018 11:47:37 AM hudson.WebAppMain contextInitialized
SEVERE: Failed to initialize Jenkins
java.lang.ExceptionInInitializerError
	at jenkins.model.Jenkins.<clinit>(Jenkins.java:5141)
	at hudson.WebAppMain.installLogger(WebAppMain.java:294)
	at hudson.WebAppMain.contextInitialized(WebAppMain.java:128)
	at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:952)
	at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:558)
	at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:917)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:370)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1497)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1459)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:847)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:287)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:545)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:138)
	at org.eclipse.jetty.server.Server.start(Server.java:416)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:108)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:113)
	at org.eclipse.jetty.server.Server.doStart(Server.java:383)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at winstone.Launcher.<init>(Launcher.java:184)
	at winstone.Launcher.main(Launcher.java:354)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at Main._main(Main.java:344)
	at Main.main(Main.java:160)
Caused by: java.lang.IllegalStateException: Expected 1 instance of hudson.security.PermissionRegistry but got 0
	at hudson.ExtensionList.lookupSingleton(ExtensionList.java:450)
	at hudson.security.PermissionRegistry.getInstance(PermissionRegistry.java:50)
	at hudson.security.PermissionGroup.register(PermissionGroup.java:161)
	at hudson.security.PermissionGroup.<init>(PermissionGroup.java:78)
	at hudson.security.PermissionGroup.<init>(PermissionGroup.java:64)
	at hudson.security.Permission.<clinit>(Permission.java:297)
	... 27 more
```

Do I need to self-register? I changed the class to extend `ExtensionPoint` and annotated it with `@Extension`, but I'm not sure if it's loaded yet by the time the Jenkins class is being loaded.

jvz replied:
Doesn't seem like it's possible to use an `ExtensionList` for an extension required by `Jenkins` due to initialization order. This could potentially use a MetaInf type service or something else lower level that executes before `Jenkins` is initialized.

jglick commented on core/src/main/java/hudson/security/PermissionRegistry.java:
```diff
@@ -0,0 +1,85 @@
+ */
+@ParametersAreNonnullByDefault
+public enum PermissionRegistry {
+    INSTANCE;
```
and this should not be publicly accessible

jglick reviewed (commented):
Without commenting on the details of whether the code in this PR is technically correct or not, I have to ask what the broader purpose of this change is. Presumably you want to move to a world in which some Jenkins installations indeed grant `ROOT` ~ `RUN_SCRIPTS` + `UPLOAD_PLUGINS` + `CONFIGURE_UPDATECENTER` to no one (just `SYSTEM`) but `ADMINISTER` to some real users.

But it would take a (large) JEP to even describe what that world would look like. It was attempted in the CloudBees DEV@cloud service, and has been proposed for Evergreen (CC @batmat)—in both cases you want to turn Jenkins into a consumer appliance, basically like an unrooted phone. This presupposes that

* You have a strictly controlled set of plugins offered on your (presumably custom) update center or whatever other distribution mechanism.
* You are using some combination of `ExtensionFilter`s and `PluginServletFilter` registrations to suppress every chunk of UI or REST endpoint or CLI command which offers to let users do things like edit and save `/configureSecurity/`.
* Every check for `ADMINISTER` in core and these plugins has been personally vetted to make sure that `ROOT` (or `RUN_SCRIPTS`) was not meant instead, or that this code path is not invokable by users due to the above filters.

To take just one example of what I suppose would be hundreds, if `configuration-as-code` is installed, then [this](https://github.com/jenkinsci/configuration-as-code-plugin/blob/d9171d086fbbf92cb5807da2dd12d6fc43df7a6b/plugin/src/main/java/io/jenkins/plugins/casc/cli/ApplyConfigurationCommand.java#L28) would need to be patched first, lest you

```yaml
jenkins:
  authorizationStrategy:
    loggedInUsersCanDoAnything:
```

and clobber whatever carefully tailored strategy was in place before. As another little example, [this code](https://github.com/jenkinsci/jenkins/blob/4d384da6f794fa1936d4142ed719ea94013649bd/core/src/main/java/jenkins/model/Jenkins.java#L3742) would need to be edited lest someone change 0 to 1 and then

```groovy
node('master') {
  sh 'cat ../secrets/initialAdminPassword | mail me@x.com'
}
```

or worse.

jvz committed:
Improve docs of permissions and reduce API surface

jvz commented:
@jglick the nontrivial amount of permission checks and changes that will be required to properly support this feature is why I'm approaching it with a feature flag to opt in to. I figured this might need a JEP to properly educate everyone on proper permission usage going forward. This also allows for the introduction of more fine grained permissions, and I'm also hoping to address the issues regarding introducing new permissions and config migration strategies and such.

I mainly opened this PR to get initial feedback on my approach to see what the ramifications could be.

jvz committed:
Fix javadoc link to public field

jvz committed:
Use a more popular form of the singleton pattern

There was an attempt at making this class an ExtensionPoint, but due to initialization order, extensions are not loaded until sometime after the Jenkins class is initialized (which includes constructing some base permissions).

jvz committed:
Simplify sort

jvz committed:
Simplify permission fromId lookup

This will use an existing instance if the permission is already known by id.

jvz committed:
Fix typo in javadoc link

jvz committed:
Use ConcurrentSkipListSet instead of synchronized

jvz committed:
Rewrite permission registration to be declarative

This uses SezPoz to load permission instances lazily.

jvz committed:
Refactor PermissionRegistry into PermissionLoader for lazy loading

This also helps avoid referencing Jenkins' uberClassLoader before it's been initialized.

jvz committed:
Fix JEP-200 error

jvz committed:
Sort permissions and permission groups

jvz committed:
Add flattened pom to gitignore

jvz committed:
Add missing permissions and groups

jvz committed:
Move feature flag to Functions

kuisathaverat commented:
I like it so much, it is a challenge movement in security management. Would be the next movement to manage permission check with annotations? This could remove thousand of stupid ”if”s and makes the method/feature permission association more clear IMHO

jvz commented:
I'd love to use annotations instead, but the typical implementation of that is to specify the role name in the annotation. I tried out a stapler proxy to implement the `@RolesAllowed` annotation, but as is, that would require specifying a long name like `@RolesAllowed("hudson.model.Hudson.ConfigureUpdateCenter")`, though since most permissions start with `hudson.model.`, maybe that could be a default.

jvz committed:
Rename @GlobalPermission to @DeclarePermission

kuisathaverat commented:
It is not a drama to put the long name 😄

jvz commented:
Here's an example annotation implementation for Stapler at least:

```java
/*
 * The MIT License
 *
 * Copyright (c) 2018 CloudBees, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package hudson.security;

import hudson.model.Hudson;
import jenkins.model.Jenkins;
import org.kohsuke.stapler.Function;
import org.kohsuke.stapler.StaplerRequest;
import org.kohsuke.stapler.StaplerResponse;
import org.kohsuke.stapler.interceptor.Interceptor;
import org.kohsuke.stapler.interceptor.InterceptorAnnotation;

import javax.annotation.CheckForNull;
import javax.servlet.ServletException;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.reflect.InvocationTargetException;

@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.METHOD /*, ElementType.TYPE */})
@InterceptorAnnotation(RequirePermission.Processor.class)
public @interface RequirePermission {
    Class<?> owner() default Jenkins.class;

    String name();
    
    class Processor extends Interceptor {
        private @CheckForNull Permission requiredPermission;
        
        @Override
        public void setTarget(Function target) {
            RequirePermission annotation = target.getAnnotation(RequirePermission.class);
            if (annotation != null) {
                Class<?> owner = annotation.owner() == Jenkins.class ? Hudson.class : annotation.owner();
                PermissionGroup group = PermissionGroup.get(owner);
                if (group != null) {
                    requiredPermission = group.find(annotation.name());
                }
            }
            super.setTarget(target);
        }

        @Override
        public Object invoke(StaplerRequest request, StaplerResponse response, Object instance, Object[] arguments) throws IllegalAccessException, InvocationTargetException, ServletException {
            if (requiredPermission != null) {
                Jenkins.get().checkPermission(requiredPermission);
            }
            return target.invoke(request, response, instance, arguments);
        }
    }
}
```

jvz committed:
Merge branch 'master' into perms

daniel-beck commented:
Are you still seeking to get this merged?

FWIW it seems to add another barely used, badly supported (plugins confuse the permissions regularly) feature flag and admins wanting to use it basically need to be Jenkins maintainers to set it up properly (and probably even get tripped up then). Not convinced it's net positive.

jvz commented:
Not really at this time. Been focusing on some other key areas to improve security instead.
